pyfar/dsp/regularization.py

```diff
@@ -0,0 +1,313 @@
+            n_samples = max(signal.n_samples, target.n_samples)
+            signal = pf.dsp.pad_zeros(signal, (n_samples-signal.n_samples))
+            target = pf.dsp.pad_zeros(target, (n_samples-target.n_samples))
+
+        # get regularization factors
+        regu = self.regularization(signal)
+
+        # Apply optional normalization
+        if normalize_regularization is not None:
+            normalization_factor_ = \
+                self.normalization_factor(signal, normalize_regularization)
+            regu.freq *= normalization_factor_
+
+        data = signal.freq
+
+        # calculate inverse filter
+        inverse = signal.copy()
+        inverse.freq = \
+            np.conj(data) / (np.conj(data) * data + beta *
+                             np.abs(regu.freq)**2)
+
+        # Apply target function
+        if target is not None:
+            inverse.freq *= target.freq
+
+        return inverse
+
+
+    def regularization(self, signal):
+        r"""
+        Get the frequency dependent regularization factors :math:`\epsilon(f)`.
+
+        Returns the regularization factors without any normalization apllied.
+
+        Parameters
+        ----------
+        signal : pf.Signal, optional
+            Signal on which the regularization will be used on. This is
+            nescessary to assure that the regularization factors have the same
+            number of bins as the signal.
+
+        Returns
+        -------
+        regu : pf.FrequencyData
+            Frequency dependent regularization as FrequencyData.
+        """
+        # Call private method to get regularization factors
+        if self._regu_type == "frequency range":
+            regu = self._get_regularization_from_frequency_range(signal)
+        elif self._regu_type == "magnitude spectrum":
+            regu = self._get_regularization_from_regularization_signal(signal)
+
+        return regu
+
+
+    def normalization_factor(self, signal, normalize_regularization):
+        r"""Returns the normalization factor by which the regularization is
+        scaled during inversion.
+
+        Parameters
+        ----------
+        signal : pf.Signal
+            Signal on which the regularization will be used on. The
+            normalization is being calculated in relation to the signal.
+        normalize_regularization : str
+            Method of normalization.
+
+            ``'max'``
+                Normalize the regularization :math:`\epsilon` to the maximum
+                magnitude of a given signal :math:`S(f)`.
+
+                :math:`Normalization Factor = \frac{\max(|S(f)|)}{\max(|\epsilon(f)|)}`
+            ``'mean'``
+                Normalize the regularization :math:`\epsilon` to the mean
+                magnitude of a given signal :math:`S(f)`.
+
+                :math:`Normalization Factor = \frac{\text{mean}(|S(f)|)}{\text{mean}(|\epsilon(f)|)}`
+
+        Returns
+        -------
+        normalization_factor : float
```

Comment:
```suggestion
        beta : float
```

Reply:
I like the verbose parameter name. But how about mentioning the 'beta' in the docstring as math symbol?

Reply:
Do we want to treat 'beta' as an equivalent for the normalization factor. Technically in this implementation we can still adjust beta after applying normalization.